File: src/main/java/hk/qingke/learn/usermanager/service/UserCreateService.java
package hk.qingke.learn.usermanager.service;

import hk.qingke.learn.usermanager.domain.UserEntity;
import hk.qingke.learn.usermanager.service.check.UserDuplicationCheck;
import hk.qingke.learn.usermanager.service.gateway.repository.UserRepository;

public class UserCreateService {
    private final UserRepository userRepository;
    private final UserDuplicationCheck userDuplicationCheck;

    public UserCreateService(UserRepository userRepository) {
        this.userRepository = userRepository;
        this.userDuplicationCheck = new UserDuplicationCheck(userRepository);
    }


    public UserEntity create(UserEntity userEntity) {
        this.userDuplicationCheck.checkWhenCreateUser(userEntity);

        return this.userRepository.save(userEntity);
    }
}

File: service/src/main/kotlin/hk/qingke/kotlin/usermanager/service/UserCreateService.kt
package hk.qingke.kotlin.usermanager.service

import hk.qingke.kotlin.usermanager.domain.UserEntity
import hk.qingke.kotlin.usermanager.service.adapter.repository.UserRepository
import hk.qingke.kotlin.usermanager.service.exception.EmailDuplicationException
import hk.qingke.kotlin.usermanager.service.exception.UsernameDuplicationException

class UserCreateService(private val userRepository: UserRepository) {

    fun create(userEntity: UserEntity): UserEntity {
        if (this.userRepository.queryUserByUsername(userEntity.username) != null) {
            throw UsernameDuplicationException()
        }

        if (this.userRepository.queryUserByEmail(userEntity.email) != null) {
            throw EmailDuplicationException()
        }

        return this.userRepository.save(userEntity)
    }

}